flow/envs/multiagent_env.py

```diff
@@ -0,0 +1,345 @@
+from copy import deepcopy
+import numpy as np
+import random
+import traceback
+from gym.spaces import Box
+
+from traci.exceptions import FatalTraCIError
+from traci.exceptions import TraCIException
+
+from ray.rllib.env import MultiAgentEnv
+
+from flow.envs.base_env import Env
+
+
+class MultiEnv(MultiAgentEnv, Env):
+    """Multi-agent version of base env. See parent class for info"""
+
+    def step(self, rl_actions):
+        """Advance the environment by one step.
+
+        Assigns actions to autonomous and human-driven agents (i.e. vehicles,
+        traffic lights, etc...). Actions that are not assigned are left to the
+        control of the simulator. The actions are then used to advance the
+        simulator by the number of time steps requested per environment step.
+
+        Results from the simulations are processed through various classes,
+        such as the Vehicles and TrafficLights classes, to produce standardized
+        methods for identifying specific network state features. Finally,
+        results from the simulator are used to generate appropriate
+        observations.
+
+        Parameters
+        ----------
+        rl_actions: numpy ndarray
+            an list of actions provided by the rl algorithm
+
+        Returns
+        -------
+        observation: dict of numpy ndarrays
+            agent's observation of the current environment
+        reward: dict of floats
+            amount of reward associated with the previous state/action pair
+        done: dict of bools
+            indicates whether the episode has ended
+        info: dict
+            contains other diagnostic information from the previous action
+        """
+        for _ in range(self.env_params.sims_per_step):
```

Comment:
is this just duplicated from changes in `base_env.py`? if so, should we delete one?

Reply:
So; it's not just duplicated, there is a slightly different part at the bottom. It's mostly duplicated and I can't figure out how to not reuse this code.

Reply:
kk, we still will need a tutorial or something that describes when to use what